File: c++/24. Swap Nodes in Pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if(head == NULL) return NULL;
        vector<int> swap;
        while(head){
            swap.push_back(head->val);
            head = head->next;
        }
        for(int i=0;i<swap.size()-1;i+=2){
            int t1 = swap[i];
            int t2 = swap[i+1];
            swap[i] = t2;
            swap[i+1] = t1;
        }
        ListNode* start = NULL;
        ListNode* end = NULL;
        for(auto x : swap){
            if(start == NULL){
                start = new ListNode(x);
                end = start;
            }
            else{
                end->next = new ListNode(x);
                end = end->next;
            }
        }
        return start;
    }
};

File: python/24. Swap Nodes in Pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head

        dummy = ListNode(0)
        dummy.next = head
        head2 = dummy

        while head2.next and head2.next.next:
            l = head2.next
            r = head2.next.next
            head2.next = r
            l.next = r.next
            r.next = l
            head2 = head2.next.next

        return dummy.next